keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ImageResolver.kt

```diff
@@ -72,47 +84,65 @@ class ImageResolver(
       ) ?: throw NoImageSatisfiesConstraints(artifact.name, environment.name)
       val image = imageService.getLatestNamedImage(
         appVersion = AppVersion.parseName(artifactVersion),
-        account = account,
-        region = region
+        account = account
       ) ?: throw NoImageFound(artifactVersion)
 
-      val imageId = image.amis[region]?.first()
-        ?.also {
-          log.info("Image found for {}: {}", artifactVersion, image)
-        } ?: throw NoImageFoundForRegion(artifactVersion, region)
-      imageId to artifactVersion
+      ResolvedImages(
+        artifactVersion,
+        image.toResolvedImage(resource.spec.locations.regions.map(ClusterRegion::region))
+      )
     } else {
       val image = imageService.getLatestNamedImage(
         packageName = artifact.name,
-        account = account,
-        region = region
+        account = account
       ) ?: throw NoImageFound(artifact.name)
 
-      val imageId = (image.amis[region]?.first()
-        ?.also {
-          log.info("Image found for {}: {}", artifact.name, image)
-        }
-        ?: throw NoImageFoundForRegion(artifact.name, region))
-      imageId to image.appVersion
+      ResolvedImages(
+        image.appVersion,
+        image.toResolvedImage(resource.spec.locations.regions.map(ClusterRegion::region))
+      )
     }
   }
 
-  private suspend fun resolveFromJenkinsJob(resource: Resource<ServerGroupSpec>, imageProvider: JenkinsImageProvider): Pair<String, String> {
-    val region = resource.spec.location.region
-    val namedImage = imageService.getNamedImageFromJenkinsInfo(
+  private suspend fun resolveFromJenkinsJob(
+    resource: Resource<ClusterSpec>,
+    imageProvider: JenkinsImageProvider
+  ): ResolvedImages {
+    val image = imageService.getNamedImageFromJenkinsInfo(
       imageProvider.packageName,
       dynamicConfigService.getConfig("images.default-account", "test"),
       imageProvider.buildHost,
       imageProvider.buildName,
       imageProvider.buildNumber
     ) ?: throw NoImageFound(imageProvider.packageName)
 
-    log.info("Image found for {}: {}", imageProvider.packageName, namedImage)
-    val amis = namedImage.amis[region]
-      ?: throw NoImageFoundForRegion(imageProvider.packageName, region)
-    return amis.first() to namedImage.appVersion // todo eb: when are there multiple?
+    log.info("Image found for {}: {}", imageProvider.packageName, image)
+    return ResolvedImages(
+      image.appVersion,
+      image.toResolvedImage(resource.spec.locations.regions.map(ClusterRegion::region))
+    )
   }
 
+  private fun NamedImage.toResolvedImage(regions: Collection<String>) =
+    amis
+      .filterKeys { it in regions }
+      .mapValues { (_, amis) -> amis?.first() }
+      .filterNotNullValues()
+      .also {
+        if (it.size < regions.size) {
+          throw NoImageFoundForRegions(imageName, regions - it.keys)
+        }
```

Comment:
Detect once if we are missing an image for one or more regions rather than potentially throwing an exception for the first detected missing image and then failing to detect any further ones.